# Solution 00: .str accessor

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/raw/Bee Colony Census Data by County.csv")

In [3]:
df.head()

Year      Period    State  State ANSI Ag District  Ag District Code  \
0  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   
1  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   
2  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   
3  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   
4  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   

    County  County ANSI Value CV (%)  
0  AUTAUGA          1.0   119   27.7  
1   DALLAS         47.0    65   27.7  
2   ELMORE         51.0   190   27.7  
3   GREENE         63.0    14   27.7  
4     HALE         65.0    10   27.7

In [4]:
df['Value'].describe()

count     7830
unique    1406
top        (D)
freq      2306
Name: Value, dtype: object

The top value is (D) ????
Most of the time you'll be working with well documented (hopefully!) so you can just read the data dictionary and find out what (D) stands for.
Here's [this dataset's](../data/documents/glossary-for-bee-stats.pdf).

Now that you know what the problem is you can address it.

Let's see how many rows have '(D)' as their 'Value'

In [5]:
df[df['Value'] == '(D)']

Empty DataFrame
Columns: [Year, Period, State, State ANSI, Ag District, Ag District Code, County, County ANSI, Value, CV (%)]
Index: []

None?
let's try something different.

In [8]:
df[df['Value'].str.contains('(D)')]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  """Entry point for launching an IPython kernel.


Year      Period      State  State ANSI                  Ag District  \
5     2012  END OF DEC    ALABAMA           1                   BLACK BELT   
7     2012  END OF DEC    ALABAMA           1                   BLACK BELT   
8     2012  END OF DEC    ALABAMA           1                   BLACK BELT   
9     2012  END OF DEC    ALABAMA           1                   BLACK BELT   
19    2012  END OF DEC    ALABAMA           1  COASTAL PLAINS & GULF COAST   
30    2012  END OF DEC    ALABAMA           1              NORTHERN VALLEY   
40    2012  END OF DEC    ALABAMA           1      UPPER PLAINS & PIEDMONT   
44    2012  END OF DEC    ALABAMA           1      UPPER PLAINS & PIEDMONT   
52    2012  END OF DEC    ALABAMA           1      UPPER PLAINS & PIEDMONT   
69    2012  END OF DEC    ARIZONA           4                     NORTHERN   
73    2012  END OF DEC    ARIZONA           4                     NORTHERN   
76    2012  END OF DEC    ARIZONA           4                     SOUTHERN   
85    2012  END OF DEC   ARKANSAS           5                      CENTRAL   
89    2012  END OF DEC   ARKANSAS           5                      CENTRAL   
90    2012  END OF DEC   ARKANSAS           5                 EAST CENTRAL   
91    2012  END OF DEC   ARKANSAS           5                 EAST CENTRAL   
92    2012  END OF DEC   ARKANSAS           5                 EAST CENTRAL   
94    2012  END OF DEC   ARKANSAS           5                 EAST CENTRAL   
95    2012  END OF DEC   ARKANSAS           5                 EAST CENTRAL   
96    2012  END OF DEC   ARKANSAS           5                 EAST CENTRAL   
97    2012  END OF DEC   ARKANSAS           5                 EAST CENTRAL   
99    2012  END OF DEC   ARKANSAS           5                NORTH CENTRAL   
104   2012  END OF DEC   ARKANSAS           5                NORTH CENTRAL   
106   2012  END OF DEC   ARKANSAS           5                NORTH CENTRAL   
108   2012  END OF DEC   ARKANSAS           5                    NORTHEAST   
109   2012  END OF DEC   ARKANSAS           5                    NORTHEAST   
110   2012  END OF DEC   ARKANSAS           5                    NORTHEAST   
111   2012  END OF DEC   ARKANSAS           5                    NORTHEAST   
114   2012  END OF DEC   ARKANSAS           5                    NORTHEAST   
115   2012  END OF DEC   ARKANSAS           5                    NORTHEAST   
...    ...         ...        ...         ...                          ...   
7765  2002  END OF DEC  WISCONSIN          55                NORTH CENTRAL   
7766  2002  END OF DEC  WISCONSIN          55                NORTH CENTRAL   
7768  2002  END OF DEC  WISCONSIN          55                    NORTHEAST   
7771  2002  END OF DEC  WISCONSIN          55                    NORTHEAST   
7776  2002  END OF DEC  WISCONSIN          55                    NORTHWEST   
7779  2002  END OF DEC  WISCONSIN          55                    NORTHWEST   
7780  2002  END OF DEC  WISCONSIN          55                    NORTHWEST   
7784  2002  END OF DEC  WISCONSIN          55                SOUTH CENTRAL   
7788  2002  END OF DEC  WISCONSIN          55                    SOUTHEAST   
7789  2002  END OF DEC  WISCONSIN          55                    SOUTHEAST   
7790  2002  END OF DEC  WISCONSIN          55                    SOUTHEAST   
7794  2002  END OF DEC  WISCONSIN          55                    SOUTHEAST   
7796  2002  END OF DEC  WISCONSIN          55                    SOUTHWEST   
7803  2002  END OF DEC  WISCONSIN          55                 WEST CENTRAL   
7811  2002  END OF DEC    WYOMING          56                    NORTHEAST   
7812  2002  END OF DEC    WYOMING          56                    NORTHEAST   
7814  2002  END OF DEC    WYOMING          56                    NORTHEAST   
7815  2002  END OF DEC    WYOMING          56                    NORTHWEST   
7817  2002  END OF DEC    WYOMING          56                    NORTHWEST   
7818  2002  END OF DEC    WY

So 2306 rows have '(D)' as their 'Value' but they are not '(D)'.

Let's take a closer look at one.

In [12]:
df.iloc[5] # so this grabs the row at index 5, which we saw is in those that have (D) as 'Value'

Year                      2012
Period              END OF DEC
State                  ALABAMA
State ANSI                   1
Ag District         BLACK BELT
Ag District Code            40
County                 LOWNDES
County ANSI                 85
Value                      (D)
CV (%)                     (D)
Name: 5, dtype: object

In [13]:
df.iloc[5]['Value'] # this grabs the 'Value' column of the row at index 5

' (D)'

![reaction](https://i.imgflip.com/wwnet.jpg)

In [19]:
df.loc[df['Value'] == ' (D)', 'Value'] = 0

The two parts of `.loc[]` are 1) the _row indexer_ and 2) the _column indexer_. <br>
The _row indexer_ here is `df['Value'] == ' (D)'` which means "grab all the rows were `Value` == ' (D)' <br>
The _column indexer_ is just `Value` because we want to grab the whole __series__ `Value`.

Now you can set the (D) values to 0, but that'd mess up your math. What you really want to set them to is `Null` and `pandas` has a special way of denoting `Null`: `nan`. `NaN` means _Not a Number_ but you can't just write 'nan' as the value. To assign `nan` values to anything with `pandas` you use `pd.np.nan` which means "from `pandas` grab `numpy` from `numpy` grab `nan`."

`NumPy` is another library on top of which `pandas` is built. It's actually kinda great, lots of science depends on it check out [numpy.org](numpy.org)

In [21]:
df['Value'].astype(float)

ValueError: could not convert string to float: '10,012'

The error is that you cannot convert "10,012" to 10012 because of the comma. <br>
But there's an easy way to `.replace` a character in a string.

In [22]:
df['Value'].str.replace(',','') # look up .replace in python.

0         119
1          65
2         190
3          14
4          10
5         NaN
6          22
7         NaN
8         NaN
9         NaN
10         46
11        256
12        147
13         63
14         45
15         10
16         77
17        721
18         90
19        NaN
20        470
21        207
22         70
23        128
24        219
25        379
26        469
27        568
28         69
29         56
        ...  
7800      966
7801     3193
7802       24
7803      NaN
7804      174
7805       66
7806       88
7807      467
7808     1541
7809      281
7810     1995
7811      NaN
7812      NaN
7813     4382
7814      NaN
7815      NaN
7816    10012
7817      NaN
7818      NaN
7819        6
7820      NaN
7821      NaN
7822      NaN
7823      NaN
7824      NaN
7825      NaN
7826      NaN
7827      NaN
7828      NaN
7829      NaN
Name: Value, Length: 7830, dtype: object

In [23]:
df['Value'] = df['Value'].str.replace(',','')

df['Value'].astype(float) # Voilà

0         119.0
1          65.0
2         190.0
3          14.0
4          10.0
5           NaN
6          22.0
7           NaN
8           NaN
9           NaN
10         46.0
11        256.0
12        147.0
13         63.0
14         45.0
15         10.0
16         77.0
17        721.0
18         90.0
19          NaN
20        470.0
21        207.0
22         70.0
23        128.0
24        219.0
25        379.0
26        469.0
27        568.0
28         69.0
29         56.0
         ...   
7800      966.0
7801     3193.0
7802       24.0
7803        NaN
7804      174.0
7805       66.0
7806       88.0
7807      467.0
7808     1541.0
7809      281.0
7810     1995.0
7811        NaN
7812        NaN
7813     4382.0
7814        NaN
7815        NaN
7816    10012.0
7817        NaN
7818        NaN
7819        6.0
7820        NaN
7821        NaN
7822        NaN
7823        NaN
7824        NaN
7825        NaN
7826        NaN
7827        NaN
7828        NaN
7829        NaN
Name: Value, Length: 783

In [24]:
df['Value'] = df['Value'].astype(float)

Now you can do math.

In [25]:
df.groupby(['Year', 'State'])['Value'].mean()

Year  State         
2002  ALABAMA             517.500000
      ALASKA               54.000000
      ARIZONA            4578.375000
      ARKANSAS            509.595745
      CALIFORNIA        10577.767442
      COLORADO            587.190476
      CONNECTICUT         508.875000
      DELAWARE            150.666667
      FLORIDA            3355.188679
      GEORGIA             605.241379
      HAWAII             3591.250000
      IDAHO              5439.666667
      ILLINOIS            101.025000
      INDIANA             122.207547
      IOWA                333.714286
      KANSAS              275.041667
      KENTUCKY             66.047059
      LOUISIANA           617.296296
      MAINE               410.416667
      MARYLAND            209.764706
      MASSACHUSETTS       295.600000
      MICHIGAN           1322.918367
      MINNESOTA          1726.256410
      MISSISSIPPI         196.416667
      MISSOURI            137.972603
      MONTANA            2772.461538
      NEBRASKA   

***
***

## Solution 01: just ignore it.

You can also just ignore anything that `contains` '(D)'

In [11]:
dff = df[df['Value'].str.contains('(D)') != True]

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: UserWarning: This pattern has match groups. To actually get the groups, use str.extract.
  """Entry point for launching an IPython kernel.


In [13]:
dff['Value'] = dff['Value'].str.replace(',','')

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [14]:
dff['Value'] = dff['Value'].astype(float)

/opt/conda/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [15]:
dff.head()

Year      Period    State  State ANSI Ag District  Ag District Code  \
0  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   
1  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   
2  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   
3  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   
4  2012  END OF DEC  ALABAMA           1  BLACK BELT                40   

    County  County ANSI  Value CV (%)  
0  AUTAUGA          1.0  119.0   27.7  
1   DALLAS         47.0   65.0   27.7  
2   ELMORE         51.0  190.0   27.7  
3   GREENE         63.0   14.0   27.7  
4     HALE         65.0   10.0   27.7

In [16]:
dff[dff['State'] == 'CALIFORNIA'].groupby('Ag District')['Value'].mean()

Ag District
CENTRAL COAST           6502.500000
NORTHEAST                176.250000
NORTHERN COAST           577.000000
SACRAMENTO VALLEY      14212.192308
SAN JOAQUIN VALLEY     38645.250000
SIERRA MOUNTAINS        1905.117647
SISKIYOU-SHASTA         9837.444444
SOUTHERN CALIFORNIA    20491.761905
Name: Value, dtype: float64

This one is filtering the __dataframe__ `dff` first. From there, you `groupby` 'Ag District' and you grab the 'Value' column, calculate the mean for each group.

In [17]:
dff[dff['State'] == 'CALIFORNIA'].groupby(['Year','Ag District'])['Value'].mean()

Year  Ag District        
2002  CENTRAL COAST           1641.250000
      NORTHEAST                 87.000000
      NORTHERN COAST          1519.000000
      SACRAMENTO VALLEY      12083.333333
      SAN JOAQUIN VALLEY     24206.500000
      SIERRA MOUNTAINS        1669.200000
      SISKIYOU-SHASTA         7879.000000
      SOUTHERN CALIFORNIA    14886.285714
2007  CENTRAL COAST           5589.333333
      NORTHEAST                  6.500000
      NORTHERN COAST           314.666667
      SACRAMENTO VALLEY      11644.375000
      SAN JOAQUIN VALLEY     39020.875000
      SIERRA MOUNTAINS        1277.400000
      SISKIYOU-SHASTA         9000.000000
      SOUTHERN CALIFORNIA    19225.000000
2012  CENTRAL COAST          10785.090909
      NORTHEAST                605.000000
      NORTHERN COAST           211.333333
      SACRAMENTO VALLEY      18623.555556
      SAN JOAQUIN VALLEY     52708.375000
      SIERRA MOUNTAINS        2522.000000
      SISKIYOU-SHASTA        12633.333333
      SO

***
[Next notebook](04_index.ipynb)